library/threading.c

```diff
@@ -119,6 +119,102 @@ int (*mbedtls_mutex_unlock)( mbedtls_threading_mutex_t * ) = threading_mutex_unl
 
 #endif /* MBEDTLS_THREADING_PTHREAD */
 
+void mbedtls_rwlock_init( mbedtls_threading_rwlock_t *lock )
+{
+    if( lock == NULL || lock->is_valid )
+        return;
+
+    mbedtls_mutex_init( &lock->readers_mutex );
+    mbedtls_mutex_init( &lock->writer_mutex );
+    lock->num_readers = 0;
+    if( lock->readers_mutex.is_valid && lock->writer_mutex.is_valid )
+        lock->is_valid = 1;
+}
+
+void mbedtls_rwlock_free( mbedtls_threading_rwlock_t *lock )
+{
+    if( lock == NULL || ! lock->is_valid )
+        return;
+
+    mbedtls_mutex_free( &lock->readers_mutex );
+    mbedtls_mutex_free( &lock->writer_mutex );
+    lock->is_valid = 0;
+}
+
+int mbedtls_rwlock_lock_reader( mbedtls_threading_rwlock_t *lock )
+{
+    int status = MBEDTLS_ERR_THREADING_MUTEX_ERROR;
+
+    if( lock == NULL )
+        return (MBEDTLS_ERR_THREADING_BAD_INPUT_DATA);
+    if( !lock->is_valid )
+        return (MBEDTLS_ERR_THREADING_MUTEX_ERROR);
+
+    if( (status = mbedtls_mutex_lock( &lock->readers_mutex ) != 0 ))
+        return status;
+
+    lock->num_readers++;
+    if( lock->num_readers == 1 )
+    {
+        if( (status = mbedtls_mutex_lock( &lock->writer_mutex ) != 0 ))
+        {
+            lock->num_readers--;
+            mbedtls_mutex_unlock( &lock->readers_mutex );
+            return status;
+        }
+    }
+    status = mbedtls_mutex_unlock( &lock->readers_mutex );
+
+    return status;
+}
+
+int mbedtls_rwlock_unlock_reader( mbedtls_threading_rwlock_t *lock )
+{
+    int status = MBEDTLS_ERR_THREADING_MUTEX_ERROR;
+
+    if( lock == NULL )
+        return (MBEDTLS_ERR_THREADING_BAD_INPUT_DATA);
+    if( !lock->is_valid )
+        return (MBEDTLS_ERR_THREADING_MUTEX_ERROR);
+
+    if( (status = mbedtls_mutex_lock( &lock->readers_mutex ) != 0 ))
+        return status;
+
+    lock->num_readers--;
+    if( lock->num_readers == 0 )
+    {
+        if( (status = mbedtls_mutex_unlock( &lock->writer_mutex ) != 0 ))
```

Comment:
My understanding is that this isn't properly supported by pthreads? See

https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_mutex_lock.html

ie if another thread took this lock, the unlock would be undefined / error behaviour.

Reply:
Yes. Doesn't pthread have native RW locks?

Reply:
@tom-cosgrove-arm Yes. So does C11. So does Windows. They're all different, so if we start using reader-writer locks, we'll have to add an abstraction layer which users/integrators will have to implement. [Mbed OS](https://os.mbed.com/docs/mbed-os/v6.15/apis/rtos-apis.html) doesn't have a rwlock API (however it has condition variables and semaphores).